move docker-compose to docker compose

mrjones-plip opened the pull request.

mrjones-plip committed:
move docker-compose to docker compose

paulpascal reviewed (approved):
Straight 🙌

PS: Don't we need to update the [ci.yml](https://github.com/medic/cht-couch2pg/blob/main/.github/workflows/ci.yml) file as well and replace `docker-compose` with `docker compose` ? (L30, L33, L36)

mrjones-plip committed:
update ci to use docker compose instead of docker-compose

mrjones-plip commented:
Good catch on the CI!  I've submitted a change (see [55069b4](https://github.com/medic/cht-couch2pg/pull/149/commits/55069b4fa9d4a7317ef20dc99cf977dee557c945)) and I'll try and get CI to run, but seeing [some errors](https://github.com/medic/cht-couch2pg/actions/runs/13118818594/job/36599949616) just now:

```
 9e37bf358a5d Pull complete 
 c9fb4b7db943 Extracting [==================================================>]     394B/394B
 c9fb4b7db943 Extracting [==================================================>]     394B/394B
 c9fb4b7db943 Pull complete 
 postgres Pulled 
 Network cht-net  Creating
 Network cht-net  Created
 Volume "cht-postgres-data"  Creating
 Volume "cht-postgres-data"  Created
 Container cht-couch2pg-couch-1  Creating
 Container cht-couch2pg-postgres-1  Creating
 Container cht-couch2pg-couch-1  Created
 Container cht-couch2pg-postgres-1  Created
 Container cht-couch2pg-couch-1  Starting
 Container cht-couch2pg-postgres-1  Starting
 Container cht-couch2pg-couch-1  Started
 Container cht-couch2pg-postgres-1  Started
/usr/local/lib/node_modules/grunt-cli/node_modules/v8flags/index.js:84
  var flags = Array.from(process.allowedNodeEnvironmentFlags);
                    ^

TypeError: Cannot convert undefined or null to object
    at Function.from (native)
    at getFlags (/usr/local/lib/node_modules/grunt-cli/node_modules/v8flags/index.js:84:21)
    at /usr/local/lib/node_modules/grunt-cli/node_modules/v8flags/index.js:142:5
    at /usr/local/lib/node_modules/grunt-cli/node_modules/v8flags/index.js:70:14
    at FSReqWrap.oncomplete (fs.js:135:15)
Error: Process completed with exit code 1.
```

mrjones-plip committed:
pin grunt-cli to 1.5.0, drop node 8 & add node 14 to CI

mrjones-plip committed:
use new couchdb image, refactor compose file, add debug to bootstrap.js

mrjones-plip commented:
yikes, the small PR turns into a big one!!  

After following [Diana's advice](https://forum.communityhealthtoolkit.org/t/keeping-cht-couch2pg-updated/4617/6), I dropped Node 8 in CI test matrix and pinned `grunt-cli` to latest which is `1.5.0`.  Rebuilding the image and testing locally with this call:

```
docker compose  -f docker-compose.test.yml build cht-couch2pg&& \
     docker compose  -f docker-compose.test.yml run cht-couch2pg grunt test
```

I then got a couch auth error:

```
====> Waiting on http://cht:cht-password@couch:5984 to appear.
{ [Error: connect ECONNREFUSED 172.20.0.2:5984
  at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1107:14)
]
  errno: 'ECONNREFUSED',
  code: 'ECONNREFUSED',
  syscall: 'connect',
  address: '172.20.0.2',
  port: 5984 }
====> Waiting on http://cht:cht-password@couch:5984 to appear.
****** ERROR: Unable to connect: http://cht:cht-password@couch:5984
```

I was unable to connect to the container externally, so I updated the compose file to expose the port and to use a modern couch version.  This worked!  Tests now showed:

```
 ✔ Container cht-couch2pg-couch-1     Running                                                                                                0.0s 
 ✔ Container cht-couch2pg-postgres-1  Running                                                                                                0.0s 
Running "eslint:src" (eslint) task

Running "mochaTest:integration" (mochaTest) task

Node Version v10.24.1
- Couch [http://cht:cht-password@couch:5984] is now available.
```

Now however I'm getting an auth error when connecting to postgres.  I note that it's using the MedicOS Postgres image from [this Dockerfile](https://github.com/medic/cht-postgres/blob/main/Dockerfile) from 4 years ago. 

This is the error I get:

```
Failed to connect to postgres server postgres://cht:cht_password@postgres:5432 { [error: password authentication failed for user "cht"
  at Connection.parseE (/app/node_modules/pg/lib/connection.js:600:48)
  at Connection.parseMessage (/app/node_modules/pg/lib/connection.js:399:19)
  at Socket.<anonymous> (/app/node_modules/pg/lib/connection.js:115:22)
  at Socket.emit (events.js:198:13)
  at Socket.EventEmitter.emit (domain.js:448:20)
  at addChunk (_stream_readable.js:288:12)
  at readableAddChunk (_stream_readable.js:269:11)
  at Socket.Readable.push (_stream_readable.js:224:10)
  at TCP.onStreamRead [as onread] (internal/stream_base_commons.js:94:17)
]
  name: 'error',
  length: 99,
  severity: 'FATAL',
  code: '28P01',
  detail: undefined,
  hint: undefined,
  position: undefined,
  internalPosition: undefined,
  internalQuery: undefined,
  where: undefined,
  schema: undefined,
  table: undefined,
  column: undefined,
  dataType: undefined,
  constraint: undefined,
  file: 'auth.c',
  line: '334',
  routine: 'auth_failed' }
```

Using `postgres://cht:cht_password@postgres:5432` I am indeed unable to connect with the `psql` command line from my workstation.

Down the rabbit hole I go!!  I suspect using modern postgres might be easier?

mrjones-plip committed:
drop pg data volume, use modern postgres

mrjones-plip commented:
As an end of the day effort, I tried bumping to latest `postgres` docker image, and I got a lot further!  I'm wondering if there's some dependencies in the test scripts that are failing?  However there's 12 errors all identical to the first one (same file an line number) - I suspect  Node version error?

```
docker compose  -f docker-compose.test.yml run cht-couch2pg grunt test                 
[+] Creating 2/2
 ✔ Container cht-couch2pg-couch-1     Created                                                                                                    0.2s 
 ✔ Container cht-couch2pg-postgres-1  Created                                                                                                    0.2s 
[+] Running 2/2
 ✔ Container cht-couch2pg-postgres-1  Started                                                                                                    0.4s 
 ✔ Container cht-couch2pg-couch-1     Started                                                                                                    0.4s 
Running "eslint:src" (eslint) task


Running "mochaTest:integration" (mochaTest) task


Node Version v10.24.1
- Couch [http://cht:cht-password@couch:5984] is now available.
====> Waiting on postgres://cht:cht_password@postgres:5432 to appear.
- Postgres [postgres://cht:cht_password@postgres:5432] is now available.
  xmlforms
    1) checks migrations
    2) checks that materialized views have the required columns

  medic-logs db replication
    3) can skip replication
    4) replicates single couch record to the right table on postgres

  replication
    5) replicates different type of couch records to pg

  medic without sentinel db replication
    6) medic replication can be skipped
    7) replicates single couch record to postgres

  medic users meta db replication
    8) can skip replication
    9) replicates single couch record to the right table on postgres
    10) checks migrations

  medic _users db replication
    11) can skip replication
    12) replicates _user record without security information

  safe
    ✓ keeps initial opts and removes passwords from returned obj
    ✓ returns valid couch/pg urls for urls without passwords


  2 passing (8s)
  12 failing

  1) xmlforms
       checks migrations:
     Uncaught Error: Unknown authenticationOk message typeMessage { name: 'authenticationOk', length: 23 }
    at Connection.parseR (node_modules/couch2pg/node_modules/postgrator/node_modules/pg/lib/connection.js:426:9)
    at Connection.parseMessage (node_modules/couch2pg/node_modules/postgrator/node_modules/pg/lib/connection.js:345:17)
    at Socket.<anonymous> (node_modules/couch2pg/node_modules/postgrator/node_modules/pg/lib/connection.js:105:22)
    at Socket.EventEmitter.emit (domain.js:448:20)
    at addChunk (_stream_readable.js:288:12)
    at readableAddChunk (_stream_readable.js:269:11)
    at Socket.Readable.push (_stream_readable.js:224:10)
    at TCP.onStreamRead [as onread] (internal/stream_base_commons.js:94:17)
  
```

mrjones-plip committed:
fix passwords, pin node and postgres

mrjones-plip commented:
OK!  After many hours of trying to get this PR to get CI to a passing state,  and after getting some second opinions - I'm going to close this PR.  It looks like CI hasn't run in `main` for a release [since 2021](https://github.com/medic/cht-couch2pg/releases/tag/3.3.0) (though I do some passing CI in [Actions tab](https://github.com/medic/cht-couch2pg/actions)) .  

My vote is that we actually hard archive this so no further changes can be made to double down on the soft archive we've already done.  I'll take this idea up [with the community](https://forum.communityhealthtoolkit.org/t/keeping-cht-couch2pg-updated/4617)!
